File: Stack/Stack.java
package Stack;

public class Stack<T> {
    
    private static class Node<T> {
        private T data;
        private Node<T> next;

        public Node(T data){
            this.data = data;
            this.next = null;
        }
    }

    private Node<T> top;
    public Stack(T data){
        this.top = new Node<T>(data);
    }
    public Stack(){
        this.top = null;
    }

    public void push(T data){
        Node<T> n = new Node<T>(data);
        n.next = this.top;
        this.top = n;
        return;
    }
    
    public T peek(){
        if(this.top == null){
            return null;
        }
        return this.top.data;
    }

    public T pop(){
        if(this.top == null){
            return null;
        }
        T n = this.top.data;
        this.top = this.top.next;
        return n;
    }

    public void print(){
        Node<T> n = this.top;
        while(n != null){
            System.out.print(n.data + " ");
            n = n.next;
        }
        System.out.println();
        return;
    }

    public boolean isEmpty(){
        return this.top == null;
    }




}

File: LeetCodePython/BasicStructures/Stack.py
#Basic Stack

class Stack:
    def __init__(self, node = None):
        if node == None:
            self.head = None
            return
        elif not isinstance (node, Node):
            node = Node(node)
        self.head = node
        
    def __iter__(self):
        n = self.head
        while n:
            yield repr(n)
            n = n.next

    def add(self, node):
        if not isinstance(node, Node):
            node = Node(node)
        node.next = self.head
        self.head = node
        return

    def pop(self):
        if self.head == None:
            return None
        node = self.head
        self.head = self.head.next
        return repr(node)

    def pop_node(self):
        if self.head == None:
            return None
        node = self.head
        self.head = self.head.next
        return node
    
    def peek(self):
        if self.head == None: return None
        return repr(self.head)
    
    def peek_node(self):
        if self.head == None: return None
        return self.head
    
    def is_empty(self):
        return self.head == None



class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        
    def __repr__(self):
        return repr(self.data)

